File: Browser/Browser_Linux/src/appconfig.cpp
#include "appconfig.h"

#include <QDir>
#include <QFileInfo>
#include <QSet>

AppConfig AppConfig::load(const QString &configPath)
{
    AppConfig fallback;
    QFileInfo info(configPath);
    QDir().mkpath(info.absolutePath());

    QFile file(configPath);
    if (!file.exists()) {
        if (file.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
            QJsonObject object;
            object.insert(QStringLiteral("url"), fallback.url);
            object.insert(QStringLiteral("defaultUrl"), fallback.defaultUrl);
            object.insert(QStringLiteral("defaultPort"), fallback.defaultPort);
            object.insert(QStringLiteral("localIndexPath"), fallback.localIndexPath);
            object.insert(QStringLiteral("kiosk"), fallback.kiosk);
            object.insert(QStringLiteral("title"), fallback.title);
            object.insert(QStringLiteral("icon"), fallback.icon);
            object.insert(QStringLiteral("allowDevTools"), fallback.allowDevTools);
            file.write(QJsonDocument(object).toJson(QJsonDocument::Indented));
            file.close();
        }
        return fallback;
    }

    if (!file.open(QIODevice::ReadOnly)) {
        return fallback;
    }

    const auto data = file.readAll();
    file.close();

    const auto doc = QJsonDocument::fromJson(data);
    if (!doc.isObject()) {
        return fallback;
    }

    const auto object = doc.object();
    AppConfig cfg;
    cfg.url = object.value(QStringLiteral("url")).toString();
    cfg.defaultUrl = object.value(QStringLiteral("defaultUrl")).toString(fallback.defaultUrl);
    cfg.defaultPort = object.value(QStringLiteral("defaultPort")).toInt(fallback.defaultPort);
    cfg.localIndexPath = object.value(QStringLiteral("localIndexPath")).toString(fallback.localIndexPath);
    cfg.kiosk = object.value(QStringLiteral("kiosk")).toBool(fallback.kiosk);
    cfg.title = object.value(QStringLiteral("title")).toString();
    cfg.icon = object.value(QStringLiteral("icon")).toString();
    cfg.allowDevTools = object.value(QStringLiteral("allowDevTools")).toBool(fallback.allowDevTools);
    return cfg;
}

QString AppConfig::effectiveTitle() const
{
    return title.trimmed().isEmpty() ? hardcodedTitle : title.trimmed();
}

QString AppConfig::resolveStartupTarget(const QString &configPath) const
{
    const QString configDirectory = QFileInfo(configPath).absolutePath();
    QString target;

    if (tryResolveConfigTarget(url, configDirectory, target)) {
        return target;
    }

    if (tryResolveConfigTarget(localIndexPath, configDirectory, target, true)) {
        return target;
    }

    return buildDefaultUrl();
}

QString AppConfig::resolveCustomIconPath(const QString &configPath) const
{
    if (icon.trimmed().isEmpty()) {
        return {};
    }

    const QString configDirectory = QFileInfo(configPath).absolutePath();
    const QString candidate = QFileInfo(QDir(configDirectory), icon.trimmed()).absoluteFilePath();
    return QFileInfo::exists(candidate) ? candidate : QString{};
}

QString AppConfig::resolveFallbackIconPath() const
{
    const QString candidate = QFileInfo(QDir(QCoreApplication::applicationDirPath()), hardcodedIconRelativePath).absoluteFilePath();
    return QFileInfo::exists(candidate) ? candidate : QString{};
}

bool AppConfig::tryResolveConfigTarget(const QString &rawValue,
                                       const QString &configDirectory,
                                       QString &target,
                                       bool requireExistingFile)
{
    target.clear();
    const QString trimmed = rawValue.trimmed();
    if (trimmed.isEmpty()) {
        return false;
    }

    const QUrl asUrl(trimmed);
    if (asUrl.isValid() && asUrl.isRelative() == false && !asUrl.scheme().isEmpty()) {
        if (asUrl.scheme() == QStringLiteral("file")) {
            if (requireExistingFile && !QFileInfo::exists(asUrl.toLocalFile())) {
                return false;
            }
        }
        target = asUrl.toString();
        return true;
    }

    const QString candidatePath = QFileInfo(QDir(configDirectory), trimmed).absoluteFilePath();
    if (QFileInfo::exists(candidatePath)) {
        target = QUrl::fromLocalFile(candidatePath).toString();
        return true;
    }

    if (requireExistingFile) {
        return false;
    }

    QString normalizedWebUrl;
    if (tryNormalizeWebUrl(trimmed, normalizedWebUrl)) {
        target = normalizedWebUrl;
        return true;
    }

    return false;
}

QString AppConfig::buildDefaultUrl() const
{
    QString normalizedBaseUrl;
    if (!tryNormalizeWebUrl(defaultUrl.trimmed().isEmpty() ? QStringLiteral("http://127.0.0.1") : defaultUrl.trimmed(), normalizedBaseUrl)) {
        normalizedBaseUrl = QStringLiteral("http://127.0.0.1");
    }

    QUrl urlValue(normalizedBaseUrl);
    if (defaultPort > 0) {
        urlValue.setPort(defaultPort);
    }
    return urlValue.toString();
}

bool AppConfig::tryNormalizeWebUrl(const QString &rawValue, QString &normalizedUrl)
{
    normalizedUrl.clear();
    const QString value = rawValue.trimmed();
    if (value.isEmpty()) {
        return false;
    }

    const QUrl absoluteUrl(value);
    if (absoluteUrl.isValid() && !absoluteUrl.isRelative() &&
        (absoluteUrl.scheme() == QStringLiteral("http") || absoluteUrl.scheme() == QStringLiteral("https"))) {
        normalizedUrl = absoluteUrl.toString();
        return true;
    }

    if (!looksLikeHostOrIp(value)) {
        return false;
    }

    const QUrl withScheme(QStringLiteral("http://") + value);
    if (withScheme.isValid() && !withScheme.isRelative()) {
        normalizedUrl = withScheme.toString();
        return true;
    }

    return false;
}

bool AppConfig::looksLikeHostOrIp(const QString &value)
{
    if (value.startsWith(QStringLiteral("http://"), Qt::CaseInsensitive) ||
        value.startsWith(QStringLiteral("https://"), Qt::CaseInsensitive) ||
        value.startsWith(QStringLiteral("file://"), Qt::CaseInsensitive)) {
        return true;
    }

    if (value.contains(u'\\')) {
        return false;
    }

    if (!value.contains(u'/') && hasLikelyLocalFileExtension(value)) {
        return false;
    }

    if (value.startsWith(QStringLiteral("localhost"), Qt::CaseInsensitive)) {
        return true;
    }

    if (!value.isEmpty() && value.front().isDigit()) {
        return true;
    }

    if (value.startsWith(u'[')) {
        return true;
    }

    const QString hostPortPath = value.section(u'/', 0, 0);
    const QString hostOnly = hostPortPath.section(u':', 0, 0);
    return hostOnly.contains(u'.') && !hostOnly.contains(u' ');
}

bool AppConfig::hasLikelyLocalFileExtension(const QString &value)
{
    const QString suffix = QFileInfo(value).suffix().toLower();
    static const QSet<QString> extensions = {
        QStringLiteral("htm"),
        QStringLiteral("html"),
        QStringLiteral("php"),
        QStringLiteral("xhtml"),
        QStringLiteral("svg")
    };
    return extensions.contains(suffix);
}

File: Browser/AppConfig.cs
using System.Net;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Fotobox.WebView2Host;

public sealed class AppConfig
{
    public string? Url { get; set; }
    public string DefaultUrl { get; set; } = "http://127.0.0.1";
    public int DefaultPort { get; set; } = 8080;
    public string? LocalIndexPath { get; set; } = "wwwroot/index.html";
    public bool Kiosk { get; set; } = false;
    public string? Title { get; set; }
    public string? Icon { get; set; }
    public bool MinimizeToTray { get; set; } = true;
    public bool AllowDevTools { get; set; } = false;

    [JsonIgnore]
    public string EffectiveTitle => string.IsNullOrWhiteSpace(Title) ? Program.HardcodedTitle : Title!.Trim();

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true
    };

    public static AppConfig Load(string configPath)
    {
        if (!File.Exists(configPath))
        {
            var fallback = new AppConfig();
            Directory.CreateDirectory(Path.GetDirectoryName(configPath)!);
            File.WriteAllText(configPath, JsonSerializer.Serialize(fallback, JsonOptions));
            return fallback;
        }

        try
        {
            var json = File.ReadAllText(configPath);
            return JsonSerializer.Deserialize<AppConfig>(json, JsonOptions) ?? new AppConfig();
        }
        catch
        {
            return new AppConfig();
        }
    }

    public string ResolveStartupTarget(string configPath)
    {
        var configDirectory = Path.GetDirectoryName(configPath)!;

        if (TryResolveConfigTarget(Url, configDirectory, out var configuredTarget))
        {
            return configuredTarget;
        }

        if (TryResolveConfigTarget(LocalIndexPath, configDirectory, out var localIndexTarget, requireExistingFile: true))
        {
            return localIndexTarget;
        }

        return BuildDefaultUrl();
    }

    public string? ResolveCustomIconPath(string configPath)
    {
        if (string.IsNullOrWhiteSpace(Icon))
        {
            return null;
        }

        var configDirectory = Path.GetDirectoryName(configPath)!;
        var iconPath = Path.Combine(configDirectory, Icon.Trim());
        return File.Exists(iconPath) ? iconPath : null;
    }

    private static bool TryResolveConfigTarget(string? rawValue, string configDirectory, out string target, bool requireExistingFile = false)
    {
        target = string.Empty;
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return false;
        }

        var trimmed = rawValue.Trim();

        if (Uri.TryCreate(trimmed, UriKind.Absolute, out var absoluteUri))
        {
            if (absoluteUri.IsFile && requireExistingFile && !File.Exists(absoluteUri.LocalPath))
            {
                return false;
            }

            target = absoluteUri.AbsoluteUri;
            return true;
        }

        var candidatePath = Path.IsPathRooted(trimmed)
            ? trimmed
            : Path.GetFullPath(Path.Combine(configDirectory, trimmed));

        if (File.Exists(candidatePath))
        {
            target = new Uri(candidatePath).AbsoluteUri;
            return true;
        }

        if (requireExistingFile)
        {
            return false;
        }

        if (TryNormalizeWebUrl(trimmed, out var normalizedWebUrl))
        {
            target = normalizedWebUrl;
            return true;
        }

        return false;
    }

    private string BuildDefaultUrl()
    {
        var baseUrl = string.IsNullOrWhiteSpace(DefaultUrl) ? "http://127.0.0.1" : DefaultUrl.Trim();

        if (!TryNormalizeWebUrl(baseUrl, out var normalizedBaseUrl))
        {
            normalizedBaseUrl = "http://127.0.0.1";
        }

        var builder = new UriBuilder(normalizedBaseUrl);
        if (DefaultPort > 0)
        {
            builder.Port = DefaultPort;
        }

        return builder.Uri.ToString();
    }

    private static bool TryNormalizeWebUrl(string rawValue, out string normalizedUrl)
    {
        normalizedUrl = string.Empty;
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return false;
        }

        var value = rawValue.Trim();

        if (Uri.TryCreate(value, UriKind.Absolute, out var absoluteUri) &&
            (absoluteUri.Scheme == Uri.UriSchemeHttp || absoluteUri.Scheme == Uri.UriSchemeHttps))
        {
            normalizedUrl = absoluteUri.AbsoluteUri;
            return true;
        }

        if (!LooksLikeHostOrIp(value))
        {
            return false;
        }

        var withScheme = "http://" + value;
        if (Uri.TryCreate(withScheme, UriKind.Absolute, out var httpUri) &&
            (httpUri.Scheme == Uri.UriSchemeHttp || httpUri.Scheme == Uri.UriSchemeHttps))
        {
            normalizedUrl = httpUri.AbsoluteUri;
            return true;
        }

        return false;
    }

    private static bool LooksLikeHostOrIp(string value)
    {
        if (value.StartsWith("http://", StringComparison.OrdinalIgnoreCase) ||
            value.StartsWith("https://", StringComparison.OrdinalIgnoreCase) ||
            value.StartsWith("file://", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        if (value.Contains('\\'))
        {
            return false;
        }

        if (!value.Contains('/') && HasLikelyLocalFileExtension(value))
        {
            return false;
        }

        if (value.StartsWith("localhost", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        var hostPortPath = value.Split('/', 2)[0];
        var hostOnly = hostPortPath;

        if (hostPortPath.StartsWith('[') && hostPortPath.Contains("]"))
        {
            var endIndex = hostPortPath.IndexOf(']');
            hostOnly = hostPortPath[..(endIndex + 1)];
        }
        else if (hostPortPath.Count(c => c == ':') == 1)
        {
            var colonIndex = hostPortPath.LastIndexOf(':');
            var possiblePort = hostPortPath[(colonIndex + 1)..];
            if (int.TryParse(possiblePort, out _))
            {
                hostOnly = hostPortPath[..colonIndex];
            }
        }

        var trimmedHost = hostOnly.Trim('[', ']');
        if (IPAddress.TryParse(trimmedHost, out _))
        {
            return true;
        }

        return trimmedHost.Contains('.') &&
               trimmedHost.All(ch => char.IsLetterOrDigit(ch) || ch is '-' or '.');
    }

    private static bool HasLikelyLocalFileExtension(string value)
    {
        var fileName = Path.GetFileName(value);
        var extension = Path.GetExtension(fileName);
        return !string.IsNullOrWhiteSpace(extension);
    }
}
